### PR TITLE
[MUL-265] Vendor can delete images

### PR DESCRIPTION
**Pull request by Quiet0o**

### CODE REVIEW

File: features/vendor/vendor_can_remove_product_lisitng_image.feature [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    And I follow "Product listings"
+    And I follow "Edit"
+    Then I should see image
+    And I follow "Delete"
+    And I click "Save draft" button
+    Given I am on "/"
+    And I follow "My account"
+    And I follow "Product listings"
+    And I follow "Edit"
+    Then I should not see image
```
Comment (senghe) [inline features/vendor/vendor_can_remove_product_lisitng_image.feature:26]: The scenario should have only one list of Given/When/Then. Here we have two scenarios in one - it needs to be split.

File: src/Component/ProductListing/Repository/DraftImageRepository.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * You can find more information about us on https://bitbag.io and write us
+ * an email on hello@bitbag.io.
+ */
```
Comment (senghe) [inline src/Component/ProductListing/Repository/DraftImageRepository.php:10]:
```suggestion
/*
 * This file has been created by developers from BitBag.
 * Feel free to contact us once you face any issues or want to start
 * You can find more information about us on https://bitbag.io and write us
 * an email on hello@bitbag.io.
 */
 
declare(strict_types=1);
```

File: src/Component/ProductListing/Repository/DraftImageRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * You can find more information about us on https://bitbag.io and write us
+ * an email on hello@bitbag.io.
+ */
```
Comment (senghe) [inline src/Component/ProductListing/Repository/DraftImageRepositoryInterface.php:10]:
```suggestion
/*
 * This file has been created by developers from BitBag.
 * Feel free to contact us once you face any issues or want to start
 * You can find more information about us on https://bitbag.io and write us
 * an email on hello@bitbag.io.
 */
 
declare(strict_types=1);
```

File: src/Component/ProductListing/Repository/DraftImageRepository.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    public function findVendorDraftImagesQuery(DraftInterface $draft): QueryBuilder
+    {
+        $draftId = $draft->getId();
+
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.owner = :draft')
+            ->setParameter('draft', $draftId)
+            ;
+    }
```
Comment (senghe) [inline src/Component/ProductListing/Repository/DraftImageRepository.php:38]: There is no need for splitting this into two methods. Additionally, the query should be tested in PHPUnit

File: features/vendor/vendor_can_remove_product_lisitng_image.feature [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  Background:
+    Given there is an "verified" vendor user "vendor" with password "vendor"
+    And I am logged in as "vendor@email.com"
+    And the store operates on a channel named "Web-US" in "USD" currency
+    And there is an admin user "admin" with password "password"
+
+  @ui
+  Scenario: Trying to edit removed product listing
+    Given There is a product listing with code "product-listing-code" and name "product-listing-name" and status "verified" with attribute and image
```
Comment (senghe) [inline features/vendor/vendor_can_remove_product_lisitng_image.feature:14]:
```suggestion
    Given There is a verified product listing "product-listing-code" named "product-listing-name"
    And the product listing has an attribute
    And the product listing has an image attached
```

File: src/Component/ProductListing/ListingPersister.php [hunk 6]
```diff
@@ -67,4 +74,21 @@ public function uploadImages(
             $this->imageUploader->upload($image);
         }
     }
+
+    public function deleteImages(DraftInterface $productDraft): void
+    {
+        $productDraftImages = $this->draftRepository->findVendorDraftImages($productDraft);
+        foreach ($productDraftImages as $image) {
+            $entity = $this->entityManager->find(DraftImage::class, $image->getId());
```
Comment (senghe) [inline src/Component/ProductListing/ListingPersister.php:82]: I do not get this line

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Component/ProductListing/ListingPersister.php (hunk 6 above)
Comment (Quiet0o) [reply src/Component/ProductListing/ListingPersister.php]: It was totally unnecessary line

File: tests/Integration/Repository/DraftImageRepositoryTest.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public function test_it_finds_image_for_product_listing(): void
+    {
+        $this->loadFixturesFromFile('DraftImageRepositoryTest/test_it_finds_image_for_product_listing.yml');
+
+        $draft = $this->entityManager->getRepository(Draft::class)->findOneBy(['versionNumber' => '1']);
+        $queryBuilder = $this->repository->findVendorDraftImages($draft);
+
+        self::assertCount(1, $queryBuilder);
```
Comment (senghe) [inline tests/Integration/Repository/DraftImageRepositoryTest.php:34]:
```suggestion
        $vendorImages = $this->repository->findVendorDraftImages($draft);

        $this->assertCount(1, $vendorImages);
```

File: tests/Integration/Repository/DraftImageRepositoryTest.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+declare(strict_types=1);
+
+namespace Tests\BitBag\OpenMarketplace\Integration\Repository;
+
+use ApiTestCase\JsonApiTestCase;
+use BitBag\OpenMarketplace\Component\ProductListing\Entity\Draft;
+use BitBag\OpenMarketplace\Component\ProductListing\Entity\DraftImage;
+
+class DraftImageRepositoryTest extends JsonApiTestCase
```
Comment (senghe) [inline tests/Integration/Repository/DraftImageRepositoryTest.php:18]:
```suggestion
final class DraftImageRepositoryTest extends JsonApiTestCase
```

File: features/vendor/vendor_can_remove_product_lisitng_image.feature [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  @ui
+  Scenario: Trying to edit removed product listing
+    Given the product listing has an image attached
+    When I am on "/"
+    And I follow "My account"
+    And I follow "Product listings"
+    And I follow "Edit"
+    Then I should see image
+    And I follow "Delete"
+    And I click "Save draft" button
```
Comment (senghe) [inline features/vendor/vendor_can_remove_product_lisitng_image.feature:22]: This is not Then

File: features/vendor/vendor_can_remove_product_lisitng_image.feature [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+  @ui
+  Scenario: Visiting product listing after deleted image:
+    Given I am on "/"
+    And I follow "My account"
+    And I follow "Product listings"
+    And I follow "Edit"
```
Comment (senghe) [inline features/vendor/vendor_can_remove_product_lisitng_image.feature:29]: This is When

File: features/vendor/create_product_listing_without_price.feature [hunk 11]
```diff
@@ -9,7 +9,7 @@ Feature:Creating a product listing without price.
     And the store operates on a channel named "Web-US" in "USD" currency
 
   @ui
-  Scenario:
+  Scenario: Creating a product listing without price
     Given I am on "/"
     And I follow "My account"
     And I follow "Product listings"
```
Comment (senghe) [inline features/vendor/create_product_listing_without_price.feature:15]: This is When

File: src/Component/ProductListing/ListingPersister.php [hunk 12]
```diff
@@ -67,4 +73,20 @@ public function uploadImages(
             $this->imageUploader->upload($image);
         }
     }
+
+    public function deleteImages(DraftInterface $productDraft): void
```
Comment (senghe) [inline src/Component/ProductListing/ListingPersister.php:77]: This class should be specced.

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+use BitBag\OpenMarketplace\Component\Vendor\Entity\VendorInterface;
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\ORM\EntityManagerInterface;
+use Gaufrette\Filesystem;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Model\ImageInterface;
+use Sylius\Component\Core\Uploader\ImageUploaderInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+
+class ListingPersisterSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:31]: final :pray:

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        $this->createNewProductListing($productDraft, $vendor);
+
+        $productListing->setCode('ABC123')->shouldHaveBeenCalled();
+        $productListing->insertDraft($productDraft)->shouldHaveBeenCalled();
+        $productListing->setVendor($vendor)->shouldHaveBeenCalled();
+        $productDraft->setProductListing($productListing)->shouldHaveBeenCalled();
+        $productDraft->ownRelations()->shouldHaveBeenCalled();
+    }
+
+    public function it_resolve_latest_draft(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:81]:
```suggestion
    public function it_resolves_latest_draft(
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        ListingInterface $productListing,
+        DraftGeneratorInterface $draftGenerator,
+        DraftInterface $productDraft
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->resolveLatestDraft($productListing);
+    }
+
+    public function it_update_latest_draft_with(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:90]:
```suggestion
    public function it_updates_latest_draft_with(
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+        ListingInterface $productListing,
+        DraftInterface $productDraft,
+        DraftGeneratorInterface $draftGenerator
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->resolveLatestDraft($productListing);
+        $this->uploadImages($productDraft);
+    }
+
+    public function it_upload_images(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:100]:
```suggestion
    public function it_uploads_images(
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+        $draftImageRepository->findVendorDraftImages($productDraft)->willReturn([$draftImage->getWrappedObject()]);
+
+        $this->deleteImages($productDraft);
+
+        $entityManager->remove($draftImage)->shouldBeCalledOnce();
+        $filesystem->has($path)->shouldHaveBeenCalled();
+        $filesystem->delete($path)->shouldBeCalled();
+    }
+
+    public function it_vendor_can_not_delete_images_that_not_exist(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:128]: This is not a correct sentence

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        $draftImageRepository->findVendorDraftImages($productDraft)->willReturn([]);
+
+        $this->deleteImages($productDraft);
+
+        $entityManager->remove($draftImage)->shouldNotBeCalled();
+        $filesystem->has($path)->shouldNotBeCalled();
+        $filesystem->delete($path)->shouldNotBeCalled();
+    }
+
+    public function it_vendor_can_not_delete_images_with_incorrect_path(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:146]: It is not correct sentence

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    public function it_creates_new_product_listing(
+        DraftInterface $productDraft,
+        VendorInterface $vendor,
+        ListingInterface $productListing,
+        FactoryInterface $productListingFactory
+    ): void {
+        $productListingFactory->createNew()->willReturn($productListing);
+        $productDraft->getCode()->willReturn('ABC123');
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:70]: Please use better string constants. If it's product code, better would be to use `mug`

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    public function it_update_latest_draft_with(
+        ListingInterface $productListing,
+        DraftInterface $productDraft,
+        DraftGeneratorInterface $draftGenerator
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->resolveLatestDraft($productListing);
+        $this->uploadImages($productDraft);
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:97]: I do not get what the test here does. There is no verification what has been done. Additionally we've two methods running on the test. Is that correct?

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    public function it_resolves_latest_draft(
+        ListingInterface $productListing,
+        DraftGeneratorInterface $draftGenerator,
+        DraftInterface $productDraft
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->resolveLatestDraft($productListing);
+        $draftGenerator->generateNextDraft($productListing)->shouldBeCalledOnce();
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:88]:
```suggestion
        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
        
        $this->resolveLatestDraft($productListing);
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    public function it_updates_latest_draft_with(
+        ListingInterface $productListing,
+        DraftInterface $productDraft,
+        DraftGeneratorInterface $draftGenerator,
+        ImageUploaderInterface $imageUploader,
+        ImageInterface $image
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->uploadImages($productDraft);
+        $imageUploader->upload($image)->shouldBeCalledOnce();
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:100]:
```suggestion
        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
        $imageUploader->upload($image)->shouldBeCalledOnce();
        
        $this->uploadImages($productDraft);
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+        $path = '/path/to/your/file.jpg';
+        $draftImage->getPath()->willReturn($path);
+        $filesystem->has($path)->willReturn(true);
+        $draftImageRepository->findVendorDraftImages($productDraft)->willReturn([$draftImage->getWrappedObject()]);
+
+        $this->deleteImages($productDraft);
+
+        $entityManager->remove($draftImage)->shouldBeCalledOnce();
+        $filesystem->has($path)->shouldHaveBeenCalled();
+        $filesystem->delete($path)->shouldBeCalled();
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:128]: Why in one place we have "Should have" but in other similar we have "Should be"? Additionally, the "Should be" thingies should be placed before line of deleting images.

Comment (senghe) [reply spec/Component/ProductListing/ListingPersisterSpec.php:128]: Please correct this for all cases below.

File: tests/Behat/Context/Ui/Vendor/ProductListingContext.php [hunk 24]
```diff
@@ -151,88 +162,15 @@ private function getPage()
...
-            $this->entityManager->persist($productListing);
-            $this->entityManager->persist($productDraft);
-            $this->entityManager->persist($productTranslation);
-            $this->entityManager->persist($productPricing);
-
-            $this->sharedStorage->set('product_listing', $productListing);
-        }
-
-        $this->entityManager->flush();
+        $this->createProudctListinByVendor($count, $status);
```
Comment (senghe) [inline tests/Behat/Context/Ui/Vendor/ProductListingContext.php:173]: A typo

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    public function it_can_not_delete_images_that_not_exist(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:131]:
```suggestion
    public function it_can_not_delete_images_that_do_not_exist(
```

File: spec/Component/ProductListing/ListingPersisterSpec.php [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+        ListingInterface $productListing,
+        DraftGeneratorInterface $draftGenerator,
+        DraftInterface $productDraft
+    ): void {
+        $draftGenerator->generateNextDraft($productListing)->willReturn($productDraft);
+        $this->resolveLatestDraft($productListing);
+        $draftGenerator->generateNextDraft($productListing)->shouldBeCalledOnce();
+    }
+
+    public function it_updates_latest_draft_with(
```
Comment (senghe) [inline spec/Component/ProductListing/ListingPersisterSpec.php:91]:
```suggestion
    public function it_updates_latest_draft(
```

With what? :thinking:

**Review by senghe — CHANGES_REQUESTED** [review]

File: tests/Behat/Context/Ui/Vendor/ProductListingContext.php [hunk 27]
```diff
@@ -301,6 +285,39 @@ public function iClickOnConfirmationModal(string $label): void
...
+        $this->entityManager->persist($channel);
+        $this->entityManager->flush();
+    }
+
+    /**
+     * @When I fill form with default data
+     */
+    public function iFillFormWithDefaultData(): void
+    {
+        $page = $this->getPage();
```
Comment (senghe) [inline tests/Behat/Context/Ui/Vendor/ProductListingContext.php:310]: The `getPage()` method shouldn't exist. We should create a separate page class for this. But I know you've just copied and pasted this. So let's stay with this as is.

File: tests/Behat/Context/Ui/Vendor/ProductListingContext.php [hunk 28]
```diff
@@ -174,15 +156,17 @@ private function getPage()
      */
     public function thereIsProductListingCreatedByVendor(int $count): void
     {
-        $this->createProudctListinByVendor($count);
+        $this->createProudctListingByVendor($count);
```
Comment (senghe) [inline tests/Behat/Context/Ui/Vendor/ProductListingContext.php:159]: Still the typo :vulcan_salute:

File: tests/Behat/Context/Ui/Vendor/ProductListingContext.php [hunk 29]
```diff
@@ -174,15 +156,17 @@ private function getPage()
...
      * @Given there is :count product listing created by vendor with status :status
      */
-    public function thereIsProductListingCreatedByVendorWithStatus2($count, $status)
-    {
-        $this->createProudctListinByVendor($count, $status);
+    public function thereIsProductListingCreatedByVendorWithStatus2(
+        int $count,
+        string $status,
+    ): void {
+        $this->createProudctListingByVendor($count, $status);
```
Comment (senghe) [inline tests/Behat/Context/Ui/Vendor/ProductListingContext.php:169]: Still the typo :)

**Review by senghe — APPROVED** [review]
I'm approving this in advance. But please fix the typos :pray:

**Pull request merged**